install_provider.sh

```diff
@@ -0,0 +1,294 @@
+    read -r -p "user [default: golem]: " ERIGON_USER
+    [[ -z "$ERIGON_USER" ]] && ERIGON_USER=golem
+fi
+
+if [[ -z "$ERIGON_DATADIR" ]]; then
+    read -r -p "erigon datadir [default: /data/erigon]: " ERIGON_DATADIR
+    [[ -z "$ERIGON_DATADIR" ]] && ERIGON_DATADIR=/data/erigon
+fi
+
+if [[ -z "$ERIGON_HOSTNAME" ]]; then
+    read -r -p "hostname: " ERIGON_HOSTNAME
+fi
+
+# set ERIGON_DISABLE_SSL=y to test this script without nagging letsencrypt services
+# ERIGON_USE_SSL is an inverse of ERIGON_DISABLE_SSL.
+[[ -z "$ERIGON_DISABLE_SSL" ]] && ERIGON_USE_SSL=y
+
+apt-get update
+
+# ========= user =========
+
+# create kvm group if not exists
+if ! getent group kvm >/dev/null 2>/dev/null; then
+    # on ubuntu <= 18.04 this package setups kvm group and access to /dev/kvm
+    apt-get install -y qemu-system-common
+fi
+
+# create user if not exists
+if ! id -u "$ERIGON_USER" >/dev/null 2>/dev/null; then
+    useradd -m -s /bin/bash -G kvm -U "$ERIGON_USER" || die "failed to create user"
+fi
+
+# add user to kvm group if doesn't already belong
+if ! id -nGz "$ERIGON_USER" | grep -qzxF kvm; then
+    adduser "$ERIGON_USER" kvm
+fi
+
+# because someone might use already existing user with HOME set to /var/lib/${ERIGON_USER}
+ERIGON_USER_HOME="$(getent passwd "$ERIGON_USER" | cut -d: -f 6)"
+[[ -z "$ERIGON_USER_HOME" ]] && die "failed to get ${ERIGON_USER} home dir"
+
+# ========= nginx & certbot =========
+
+apt-get install -y acl nginx certbot apache2-utils || die "failed to install dependencies"
+
+# stop nginx to leave port 80 for certbot standalone
+systemctl stop nginx
+
+[[ -f /etc/nginx/dhparam.pem ]] || \
+    openssl dhparam -out /etc/nginx/dhparam.pem 2048
+
+if [[ "$ERIGON_USE_SSL" == y && ! -f /etc/letsencrypt/live/$ERIGON_HOSTNAME/fullchain.pem ]]; then
+    if [[ -z "$ERIGON_EMAIL" ]]; then
+        read -r -p "email (for certbot): " ERIGON_EMAIL
+    fi
+
+    certbot certonly --standalone -d "$ERIGON_HOSTNAME" --email "$ERIGON_EMAIL" -n --agree-tos --force-renewal \
+        || die "failed to get certificate"
+fi
+
+mkdir -p /etc/letsencrypt/renewal-hooks/post/
+cat >/etc/letsencrypt/renewal-hooks/post/nginx-reload.sh <<EOF
+#!/bin/bash
+nginx -t && systemctl reload nginx
+EOF
+chmod +x /etc/letsencrypt/renewal-hooks/post/nginx-reload.sh
+
+touch /etc/nginx/erigon_htpasswd
+setfacl -m "u:${ERIGON_USER}:rw" /etc/nginx/erigon_htpasswd
+
+cat >/etc/systemd/system/reload-nginx.path <<EOF
+[Unit]
+[Unit]
+Description=Watch for changes in nginx config
+After=local-fs.target
+
+[Path]
+Unit=reload-nginx.service
+PathChanged=/etc/nginx/erigon_htpasswd
+
+[Install]
+WantedBy=default.target
+EOF
+
+cat >/etc/systemd/system/reload-nginx.service <<EOF
+[Unit]
+Description=Reload nginx
+Requisite=nginx.service
+Wants=local-fs.target
+After=nginx.service
+
+[Service]
+Type=oneshot
+ExecStart=/usr/bin/systemctl reload nginx
+
+[Install]
+WantedBy=default.target
+EOF
+
+systemctl daemon-reload
+
+cat >/etc/nginx/nginx.conf <<EOF
```

Comment:
We might benefit of bringing this file outside the script as well (just for the readability and ease of further modification)

Reply:
I can move it out of the script, but I'm using here some bash variables substitution magic. I'd have to change it into some `sed` magic or something.

Reply:
Ok, lets not complicate it further - it seems to be easiest approach.